### ISSUE TITLE
upgrade resolution for retina displays

### ISSUE THREAD

**Issue by otfried** [otfried/ipe#6]
I was thinking to actually use scalable icons for larger resolutions. For instance, the Tango icons that I'm using are already available in a scalable format.

We would still need bitmaps as a fall-back for lower resolutions. The question is whether a scalable icon already looks good on a retina display.

What would be a reasonable size of a toolbar button in pixels on a retina display? I need to check if there is a good way to find the resolution of the display (or simply let the user determine the size of toolbar buttons in pref.toolbar_size, which would also help people with poor eyesight).

I may have access to retina displays in 2016, let's see then.

**Comment by m7thon** [comment]
Apart from the icons, I think it is even more important that the drawing itself is rendered in high resolution. Here is a screenshot of what a drawing looks like in ipe vs. in my pdf-viewer on my MacBook (should be raw displayed pixels, i.e., the displayed size on screen is about 50%):
 
![ipe_rendering](https://cloud.githubusercontent.com/assets/5876537/8570365/17f34668-2580-11e5-881b-e26e3717423b.jpg)
![pdf_rendering](https://cloud.githubusercontent.com/assets/5876537/8570373/1c6202de-2580-11e5-94cf-ffe4e1ac6985.jpg)

**Comment by otfried** [comment]
Very interesting.  I had no idea that this is happening.  Apparently Mac OS X lies to the application when it asks for the pixel size of the canvas.

I just found this: http://blog.qt.io/blog/2013/04/25/retina-display-support-for-mac-os-ios-and-x11/

which suggests that if you add this to the Info.plist file:
```
<key>NSPrincipalClass</key>
<string>NSApplication</string>
<key>NSHighResolutionCapable</key>
<string>True</string>
```

Mac OS will report true pixels to the application.  (This will probably cause the icons to be much too small, but I wonder if it will fix the canvas display.)

**Comment by m7thon** [comment]
Wow, after adding the above to the Info.plist file, ipe now no longer starts from the Applications menu. 
Also, I usually launch ipe from the command line anyway, which ignores the Info.plist, so this is (for me) an unsatisfactory fix.
However, it turns out that this (low-res rendering) seems to be affecting all qt applications on my system, so this is likely a qt problem.
I'll try to investigate.

**Comment by otfried** [comment]
Hmm.  Did you add it in the right place in the Info.plist file?

Apparently you can go to the Application Info Window and there should be a checkbox that tells you if it supports high resolution.

I'm somewhat confused by your claim that the Info.plist file is ignored when you start from the command line - this seems to contradict what the web says.  Have you tried moving the Ipe executable out of the bundle and starting it there?   Apparently the default (when no Info.plist file is there) would be to ALLOW high resolution.

**Comment by m7thon** [comment]
Ok, here some clarifications:
- Adding the above to the Info.plist *does* have an effect, switching to hires mode
- This can be checked by looking at the application info, where there is a checkbox "Open in low resolution" that is unchecked when adding the above to Info.plist. As you say
- Starting from the command line indeed seems to start in hires mode by default.
- OSX does some caching of the Info.plist, so changes only take effect after *copying* the whole application bundle. Kind of annoying, but ok.

However, even hires mode does *not* affect the drawing pane on my system. See the following screenshots (now all taken after selecting "Zoom->Normal size" in ipe). Note that these are displayed at about half size on my screen:
Ipe in low res mode (note that the fonts are pixelated):
![ipe_lowres](https://cloud.githubusercontent.com/assets/5876537/8573100/27859396-2590-11e5-9858-735d35a4facf.jpg)
Ipe in high res mode (fonts look nice, but drawing pane still pixelated)
![ipe_highres](https://cloud.githubusercontent.com/assets/5876537/8573101/299ffa22-2590-11e5-8b60-5385d4d10922.jpg)
Screen grab of the same test file shown in my pdf viewer at "actual size":
![pdf_grab](https://cloud.githubusercontent.com/assets/5876537/8573109/33f3c44a-2590-11e5-8118-b608e722c00b.jpg)

I don't know if this is a qt issue. I installed ipe via MacPorts, which uses Qt 4.8.7.

For testing, I also just tried with your provided Mac binary:
* This opened in high resolution mode (perhaps due to caching?)
* Drawing (lines and circles) are rendered the same
* pdflatex run gives an error and causes ipe to hang

* your provided version requires installing a complete X window environment as a prerequisite

Hope this helps to give you an idea what is going on. It would be great to get the drawing to render in high resolution as well.

**Comment by otfried** [comment]
Interesting.  Here is an experimental patch for ipecanvas_qt.cpp:

https://www.dropbox.com/s/e7ficg34kt0u1cm/retina.patch?dl=0

It simply doubles the resolution of the backing Cairo surface, and then hopes that Qt will use the full resolution when rendering to the display.

My mistake was assuming that once you enable High Resolution, that the app would actually get device pixel coordinates from the OS.  That is not the case - all coordinates are now in points, where 2 pixels are one point.   This shouldn't be hard to take care of.

Next year I may look at actually rendering using Cocoa instead of Cairo.

As for the X window environment:  Ipe uses Cairo, Freetype,  Libpng, and Libjpeg.   Apple has chosen to distribute these only with the X window environment, so that's why you have to install it.  I should probably simply include my own builds of these in the bundle.   I assume Macports simply builds their own copy of each of these.

**Comment by m7thon** [comment]
Yes, the patch seems to do the trick. Thank you! :-) See here:
![ipe_retina](https://cloud.githubusercontent.com/assets/5876537/8590139/059b1e34-261e-11e5-97b1-dbece3acc99f.jpg)
(again, this is actually displayed smaller on screen)

A couple more things to note:
- The "Zoom->Normal size" is now tiny. I manually zoomed to window to "144 ppi" to get a comparable screenshot.
- The real size (compared to a piece of A4 held to the screen) for me corresponds to "220 ppi", and that happens to be exactly my screen resolution. Is there a place that I can set this for the normal size?
- The "arrow/fill menu" (where you can click for line endings and fill properties) is still shown in low resolution.
- The little green dotted box around the latex text is now more fine, with smaller spaced "dots". I actually like it better this way.

**Comment by m7thon** [comment]
> Ipe uses Cairo, Freetype, Libpng, and Libjpeg ...
> ... I should probably simply include my own builds of these in the bundle. I assume Macports simply builds their own copy of each of these.

Yes, MacPorts does, and I had assumed these would be included in the bundle. Installing X is easy though. A window pops up informing that X is now no longer part of the system, and gives a download link to XQuartz.

**Comment by m7thon** [comment]
Your patch does cause some problems:
- the little green cross-mark that shows where you are about to draw something is no longer shown when moving the mouse (well, it is shown after clicking, or in line drawing mode)
- snapping indications are no longer shown (except eg. when drawing lines)
- the snapping distance seems to have halved, too.
- running in "low resolution mode" no longer works well:
![ipe_retina_lowres](https://cloud.githubusercontent.com/assets/5876537/8591565/765ac970-2629-11e5-8b45-466ffb153db3.jpg)
  - the grid is not rendered properly
  - the drawing looks awful

**Comment by otfried** [comment]
Well, I said it was an experiment, to see if this approach works at all (that is, if Qt is aware of the high resolution and can actually render a bitmap with full resolution).  Now that we know that I have to fill in the details...

**Comment by m7thon** [comment]
Sure ;-)

**Comment by otfried** [comment]
Here is a snapshot of Ipe 7.1.8 that has support for retina displays and should fix all the issues you found:

https://www.dropbox.com/s/f3d3ibeeeaxi1a1/ipe-7.1.8-src.tar.gz?dl=0

You will need to enable retina display by setting prefs.retina.  You should also set the other pixel-dependent preferences, as setting prefs.retina to true means that a "pixel" is half the size if is otherwise.

**Comment by m7thon** [comment]
I just tested this (briefly), and it seems to work well as far as I can tell. Here are some remaining minor issues/comments:

* In high resolution mode (regardless of the retina setting), the interface fonts (not the latex ones) are displayed in slightly different positions. In particular, when selecting colors (e.g., foreground color), the text overlaps the little color preview square. See the image for what I mean:
![ipe_fonts](https://cloud.githubusercontent.com/assets/5876537/8627274/32adb7d8-274a-11e5-8754-fde4c68d866a.jpg)
* It would be nice for people building an app bundle (with the script, MacPorts, etc.) to include a high resolution ipe app icon. The ones under src/ipe/icons and src/ipe/ipe.ico are very low resolution.
* Is there an easier was to set this than editing .../share/ipe/7.1.8/lua/prefs.lua? Can one have local prefs under ~/.ipe/prefs.lua? Is this documented somewhere? At some point it would be nice to have a "Preferences" menu item to set such things.
* I noticed you set CXX=g++, noting that only gcc is properly tested. By default on OSX for some time now, g++ is just a link to clang++. In any case, building with clang works fine. One warning you may want to fix:
```
./ipeui_qt.h:32:9: warning: 'IPEUI_QT__H' is used as a header guard here, followed by #define of a different macro [-Wheader-guard]
```
* finally, ipe sometimes gives messages such as the following when exiting or opening a new file:
```
ipe[96424:15938846] modalSession has been exited prematurely - check for a reentrant call to endModalSession:
```
I have no idea what this could mean.

**Comment by otfried** [comment]
The standard way to change the preference settings is to put a file in .ipe/ipelets/customize.lua:

http://ipe.otfried.org/manual/manual_39.html

**Comment by otfried** [comment]
The first item seems to be a Qt problem - this happens deep inside Qt, there's nothing I can do about it.  However, I plan to write a Cocoa-based UI next year, dropping Qt entirely for Mac OS, so this problem will go away.  Better icons have to wait until then or until someone contributes them.  The last item probably happens inside Qt somewhere and should also go away next year...
